packages/aztec.js/src/proof/proof.js

```diff
@@ -0,0 +1,184 @@
+const { constants, errors } = require('@aztec/dev-utils');
+const BN = require('bn.js');
+const { Enum } = require('enumify');
+const { padLeft } = require('web3-utils');
+
+const bn128 = require('../bn128');
+const Keccak = require('../keccak');
+const ProofUtils = require('./utils');
+const types = require('./types');
+
+const { AztecError } = errors;
+
+class ProofType extends Enum {}
+ProofType.initEnum(types);
+
+/**
+ * @class
+ * @class Class to create, store and pass around AZTEC proofs and related transcripts, outputs and hashes.
+ */
+class Proof {
+    /**
+     * @param {string} type one of the ProofType enum values
+     * @param {Note[]} inputNotes input AZTEC notes
+     * @param {Note[]} outputNotes output AZTEC notes
+     * @param {string} sender Ethereum address of transaction sender
+     * @param {string} publicValue public commitment being added to proof
+     * @param {string} publicOwner holder of a public token being converted
+     * @param {Object} metadata arbitrarily-sized object
+     */
+    constructor(type, inputNotes, outputNotes, sender, publicValue, publicOwner, metadata = []) {
+        this.type = type;
+        this.inputNotes = inputNotes;
+        this.m = inputNotes.length;
+        this.outputNotes = outputNotes;
+        this.notes = [...inputNotes, ...outputNotes];
+        this.sender = sender;
+        this.publicOwner = publicOwner;
+        this.metadata = metadata;
+        if (BN.isBN(publicValue)) {
+            this.publicValue = publicValue;
+        } else if (publicValue < 0) {
+            this.publicValue = bn128.curve.n.sub(new BN(-publicValue));
+        } else {
+            this.publicValue = new BN(publicValue);
+        }
+        this.validateInputs();
+        this.constructChallengeHashes();
+    }
+
+    get challengeHex() {
+        return `0x${padLeft(this.challenge.toString(16), 64)}`;
+    }
+
+    get eth() {
+        return {
+            // Ethereum will automatically prepend a bytes array with an evm word that
+            // represents the length of the bytes array
+            output: `0x${this.output.slice(0x40)}`,
```

Comment:
How come these are different by 2 characters? we should probably document why this is the case

Reply:
I described this in a document I shared on Slack.